swaybar/tray/dbusmenu.c

```diff
@@ -0,0 +1,1381 @@
+	cairo_set_source_u32(
+		shm, menu->dbusmenu->sni->tray->bar->config->colors.focused_background);
+	cairo_paint(shm);
+
+	cairo_set_operator(shm, CAIRO_OPERATOR_OVER);
+	cairo_set_source_surface(shm, recorder, -surface_x, -surface_y);
+	cairo_paint(shm);
+
+	cairo_surface_destroy(recorder);
+	cairo_destroy(cairo);
+
+	if (dbusmenu_surface->width != surface_width ||
+			dbusmenu_surface->height != surface_height) {
+		if (dbusmenu_surface->surface) {
+			wl_surface_destroy(dbusmenu_surface->surface);
+			dbusmenu_surface->surface = NULL;
+			sway_log(SWAY_DEBUG, "Destroy xdg popup");
+			xdg_popup_destroy(dbusmenu_surface->xdg_popup);
+			dbusmenu_surface->xdg_popup = NULL;
+		}
+
+		// configure & position popup surface
+		struct wl_surface *surface = wl_compositor_create_surface(bar->compositor);
+		struct xdg_surface *xdg_surface =
+			xdg_wm_base_get_xdg_surface(menu->dbusmenu->wm_base, surface);
+		struct xdg_positioner *positioner =
+			xdg_wm_base_create_positioner(menu->dbusmenu->wm_base);
+
+		struct swaybar_dbusmenu_menu_item *item =
+			find_item(!menu->parent_menu ? menu : menu->parent_menu, menu->item_id);
+		struct swaybar_output *output = menu->dbusmenu->output;
+		int x = menu->item_id == 0 ? menu->dbusmenu->x
+								   : item->hotspot.x / output->scale;
+		int y = menu->item_id == 0 ? menu->dbusmenu->y
+								   : item->hotspot.y / output->scale;
+
+		xdg_positioner_set_offset(positioner, 0, 0);
+		// Need to divide through scale because surface width/height is scaled
+		xdg_positioner_set_size(positioner, surface_width / output->scale,
+								surface_height / output->scale);
+
+		int padding = (tray->bar->config->tray_padding * output->scale) / 2;
+		if (bar->config->position & ZWLR_LAYER_SURFACE_V1_ANCHOR_TOP) { // top bar
+			xdg_positioner_set_anchor(positioner, XDG_POSITIONER_ANCHOR_BOTTOM_LEFT);
+			xdg_positioner_set_gravity(positioner, XDG_POSITIONER_GRAVITY_BOTTOM_LEFT);
+			xdg_positioner_set_anchor_rect(positioner, x, y - padding, 1, 1);
+		} else {
+			xdg_positioner_set_anchor(positioner, XDG_POSITIONER_ANCHOR_TOP_LEFT);
+			xdg_positioner_set_gravity(positioner, XDG_POSITIONER_GRAVITY_TOP_LEFT);
+			xdg_positioner_set_anchor_rect(
+			positioner, x, y + item->hotspot.height / output->scale, 1, 1);
+		}
+
+		struct xdg_popup *xdg_popup;
+		if (!menu->parent_menu) {
+			// Top level menu
+			xdg_popup = xdg_surface_get_popup(xdg_surface, NULL, positioner);
+			zwlr_layer_surface_v1_get_popup(output->layer_surface, xdg_popup);
+		} else {
+			// Nested menu
+			xdg_popup = xdg_surface_get_popup(
+					xdg_surface, menu->parent_menu->surface->xdg_surface, positioner);
+		}
+
+		xdg_popup_grab(xdg_popup, menu->dbusmenu->seat->wl_seat,
+					   menu->dbusmenu->serial);
+		xdg_popup_add_listener(xdg_popup, &xdg_popup_listener, menu);
+		xdg_surface_add_listener(xdg_surface, &xdg_surface_listener, NULL);
+		wl_surface_commit(surface);
+
+		wl_display_roundtrip(bar->display);
```

Comment:
We do not need to do a roundtrip, you can destroy the positioner immediately after use as the messages are queued.

I believe this roundtrip is why the "drawing" state was added as well, right? If so, that can be removed as well as it makes the code rather confusing.

Reply:
Correct, that makes it easier. Thanks.

Reply:
This is not solved with a roundtrip, it's solved by not rendering anything until you have received and acid the first configure event for the surface, sent when you assigned the xdg popup role.

Reply:
Thanks, I got rid of the roundtrip.